server/src/main/java/org/elasticsearch/repositories/IndexMetaDataGenerations.java

```diff
@@ -133,12 +134,19 @@ public IndexMetaDataGenerations withAddedSnapshot(
      * @return new instance without the given snapshot
      */
     public IndexMetaDataGenerations withRemovedSnapshots(Collection<SnapshotId> snapshotIds) {
-        final Map<SnapshotId, Map<IndexId, String>> updatedIndexMetaLookup = new HashMap<>(lookup);
-        updatedIndexMetaLookup.keySet().removeAll(snapshotIds);
-        final Map<String, String> updatedIndexMetaIdentifiers = new HashMap<>(identifiers);
-        updatedIndexMetaIdentifiers.keySet()
-            .removeIf(k -> updatedIndexMetaLookup.values().stream().noneMatch(identifiers -> identifiers.containsValue(k)));
-        return new IndexMetaDataGenerations(updatedIndexMetaLookup, updatedIndexMetaIdentifiers);
+        final Map<SnapshotId, Map<IndexId, String>> retainedIndexMetaLookup = lookup.entrySet()
+            .stream()
+            .filter(e -> snapshotIds.contains(e.getKey()) == false)
+            .collect(Collectors.toUnmodifiableMap(Map.Entry::getKey, Map.Entry::getValue));
+        final Set<String> retainedBlobUUIDs = retainedIndexMetaLookup.values()
```

Comment:
We did you change this to two loops now, seems the first version that I LGTMed was more efficient because we wouldn't loop `retainedIndexMetaLookup` twice?